README.md
node-installer
==============

An installer/updater for the node client.


CHANGELOG.md
## 1.0.3 (2015-07-23)

Upgrade:
  - keybase-path 0.0.16

## 1.0.2 (2015-07-01)

Upgrade:
  - Request 2.58.0

## 1.0.1 (2014-12-23)

Bugfixes:
  - Close #66
  - upgrade versions to solve dep problems

## 1.0.0 (2014-09-23)

Features:

  - Use gpg2 rather than gpg by default
     - See keybase/keybase-issues#1002
  - Check for bad versions of node
     - See keybase/keybase-issues#968
  - Upgrade to iced-utils@v0.1.22 to support args of the form `--gpg=/usr/local/bin/gpg`
     - Before you'd have to say `--gpg /usr/local/bin/gpg`, but the actual client
       was more lenient
  - New ICS/ICS-runtime split
  - Fix bad link to documentation (see issue #58)

## 0.1.22 (2014-04-29)

Features:

  - Mention where keybase was installed in install message 

## 0.1.21 (2014-04-04)

Bugfixes:

  - Support windows homedirs with spaces

## 0.1.20 (2014-04-04)

Bugfixes:

  - Simplify and make more accurate the check to see where to install.  Use `npm get prefix`.
    This is a lot more accurate and won't give false failures that prompt to install as root.
  - Retire NPM_INSTALL_PREFIX, and just use PREFIX instead.
  - Remove the engine requirement in pacakge.json so that users get a nicer error message.

## 0.1.19 (2014-04-03)

Bugfixes:

  - Add a PREFIX environment variable for installation.

## 0.1.17 (2014-03-25)

Bugfixes:

  - Address keybase/keybase-issues#349: Fix bug with `--prefix` and npm v1.2.x
  - Check version of node.js after install, to solve some ugly crashes

## 0.1.16 (2014-03-20)

Bugfixes:

  - Close keybase/keybase-issuses#294: Upgrade to gpg-wrapper@v0.0.42, which
    creates an empty trustdb.

## 0.1.15 (2014-03-19)

Bugfixes:

  - Upgrade gpg-wrapper in build; previous build was stale

## 0.1.14 (2014-03-19)

Bugfixes:

  - Documentation nit; can use --prefix on failed install, not just sudo.

## 0.1.13 (2014-03-17)

Features:
  
  - People who fail otherwise can turn on the "nuclear option" which is to 
    ignore their GPG options when doing tmp keyring operations.  Try --no-gpg-options.
  - Upgrade to gpg-wrapper v0.0.41 with the above.

## 0.1.12 (2014-03-16)

Bugfixes:

  - Back out the previous change.
  - Upgrade gpg-wrapper to v0.0.40; this fixes indexing for people who have
    the `with-fingerprint` gpg.conf option on.

## 0.1.11 (2014-03-14)

Bugfixes:

  - Close keybase/keybase-issues#190 --- some versions of GPG print the fingerprint
    of the subkey too, so we should be ok with getting >1 (let us just call it <=2) 
    fingerprint on `gpg -k --fingerprint`.

## 0.1.10 (2014-03-10)

Features:

  - Enable the https_proxy and http_proxy environment variables

## 0.1.9 (2014-03-10)

Features:

  - Enable -x/--proxy flag for proxying the install through the given proxy

Cleanups:

  - Cleanup usage message

## 0.1.8 (2014-03-05)

Bugfixes:
 
  - Close #41 -- buggy doc message
  - Close #44 -- poop out on old version of node

## 0.1.7 (2014-03-03)

Bugfixes:

  - Fix keybase/keybase-installer#39 --- do the install check for --prefix
    if --prefix was specified.

## 0.1.6 (2014-03-03)

Features:
  
  - Better administrator install message on windows
  - Allow --prefix to specify a non-standard install location (#38)
  - A welcome message.

## 0.1.5 (2014-02-20)

Features:

  - Better error messages if you lack the permissions to run npm install without root.

## 0.1.4 (2014-02-18)

Bugfixes:

  - Close #32, another bug in windows.  Using the USERPROFILE environment
  variable rather than HOMEDIR, since the former works with path concatenations,
  and the latter doesn't seem to.

## 0.1.3 (2014-02-18)

Bugfixes:

  - Upgrade to gpg-wrapper v0.0.32 to solve a bug in keydir creation

## 0.1.2 (2014-02-18)

Bugfixes:

  - More bugfixes on windows; we had a crash on keyring on a Windows 7 Pro,
    GPG 2.0.22. I think it's this bug that gpg won't index the keyring if there's not a 
    pubring.gpg file there.  Fix with an upgrade to gpg-wrapper v0.0.31.

## 0.1.1 (2014-02-17)

Bugfixes:

  - Upgrade to gpg-wrapper v0.0.30
  - Fix bug in key upgrade protocol, with setting the final key version causing a crash
  - Fix bug in key upgrade protocol for if we have the code key but not the index key
  - Fallback to a lower keyring-resident key if the most recent one failed in the above
  check. It's always safer for to use the local keys than the ones that came with the software.
  - Fix bug #31, a cleanup error in Windows.

Features:

  - Add config::set_master_ring for server-ops
  - Operational on Windows!

## 0.1.0 (2014-02-17)

Bugfixes:

  - Use `iced-spawn` rather than the spawn logic in `gpg-utils`.
  - Upgrade to `gpg-utils` to v0.0.29
  - Hopefully some fixes for windows; it is tested to work now at least once.
  - Fix bug #18; store keys to a private keyring rather than the main
    keyring, this is much safer.

Cleanups:

  - Get rid of the `getopt` here, which moved into `iced-utils`

## 0.0.11 (2014-2-14)

Bugfixes:
  
  - Look for `npm` in path and fail more gracefully
  - Allow specification of alternate `npm` if you want that.
  - Upgrade to gpg-wrapper v0.0.25, for less dependence on parsing GPG text output.
  - More debugger output for debugging problems in the wild

## 0.0.10 (2014-2-13)

Bugfixes:

  - Unbreak `set_key_version`, which should be called whenever we call `find_latest_key`

Features:

  - Inaugural Changelog!


lib/config.js
// Generated by IcedCoffeeScript 1.7.1-f
(function() {
  var AltKeyRing, Config, a_json_parse, base64u, chain, constants, fs, fullname, home, iced, iutils, keyring, kpath, log, make_esc, mkdir_p, path, prng, request, set_gpg_cmd, tmpdir, url_join, __iced_k, __iced_k_noop, _ref, _ref1, _ref2,
    __slice = [].slice;

  iced = require('iced-runtime');
  __iced_k = __iced_k_noop = function() {};

  constants = require('./constants').constants;

  fullname = require('./package').fullname;

  request = require('./request').request;

  log = require('./log');

  tmpdir = require('os').tmpdir;

  fs = require('fs');

  _ref = require('iced-error'), chain = _ref.chain, make_esc = _ref.make_esc;

  iutils = require('iced-utils');

  _ref1 = iutils.util, a_json_parse = _ref1.a_json_parse, base64u = _ref1.base64u;

  mkdir_p = iutils.fs.mkdir_p;

  prng = require('crypto').prng;

  path = require('path');

  _ref2 = require('gpg-wrapper'), set_gpg_cmd = _ref2.set_gpg_cmd, keyring = _ref2.keyring;

  AltKeyRing = keyring.AltKeyRing;

  kpath = require('keybase-path');

  url_join = function() {
    var a, args, parts, rxx, trim;
    args = 1 <= arguments.length ? __slice.call(arguments, 0) : [];
    rxx = /^(\/*)(.*?)(\/*)$/;
    trim = function(s) {
      var m;
      if ((m = s.match(rxx)) != null) {
        return m[2];
      } else {
        return s;
      }
    };
    parts = (function() {
      var _i, _len, _results;
      _results = [];
      for (_i = 0, _len = args.length; _i < _len; _i++) {
        a = args[_i];
        _results.push(trim(a));
      }
      return _results;
    })();
    return parts.join('/');
  };

  home = function(opts) {
    var x;
    x = kpath.home(opts);
    return x;
  };

  exports.Config = Config = (function() {
    function Config(argv) {
      this.argv = argv;
      this._tmpdir = null;
      this._alt_cmds = {};
      this._keyring_dir = null;
      this._actual_prefix = null;
    }

    Config.prototype.set_actual_prefix = function(p) {
      return this._actual_prefix = p;
    };

    Config.prototype.get_keyring_dir = function() {
      var d, v;
      if (this._keyring_dir == null) {
        if (!((d = this.argv.get("k", "keyring-dir")) != null)) {
          v = (home({
            array: true
          })).concat([".keybase-installer", "keyring"]);
          d = kpath.unsplit(v);
        }
        this._keyring_dir = d;
      }
      return this._keyring_dir;
    };

    Config.prototype.init_keyring = function(cb) {
      var dir, esc, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      keyring.init({
        log: log,
        get_tmp_keyring_dir: (function(_this) {
          return function() {
            return _this.get_tmpdir();
          };
        })(this),
        get_no_options: (function(_this) {
          return function() {
            return _this.get_no_gpg_options();
          };
        })(this)
      });
      dir = this.get_keyring_dir();
      esc = make_esc(cb, "Config::init_keyring");
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/node-installer/src/config.iced",
            funcname: "Config.init_keyring"
          });
          AltKeyRing.make(dir, esc(__iced_deferrals.defer({
            assign_fn: (function(__slot_1) {
              return function() {
                return __slot_1._master_ring = arguments[0];
              };
            })(_this),
            lineno: 66
          })));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          (function(__iced_k) {
            __iced_deferrals = new iced.Deferrals(__iced_k, {
              parent: ___iced_passed_deferral,
              filename: "/home/max/src/keybase/node-installer/src/config.iced",
              funcname: "Config.init_keyring"
            });
            _this._master_ring.index(esc(__iced_deferrals.defer({
              assign_fn: (function(__slot_1) {
                return function() {
                  return __slot_1._keyring_index = arguments[0];
                };
              })(_this),
              lineno: 67
            })));
            __iced_deferrals._fulfill();
          })(function() {
            return cb(null);
          });
        };
      })(this));
    };

    Config.prototype.keyring_index = function() {
      return this._keyring_index;
    };

    Config.prototype.master_ring = function() {
      return this._master_ring;
    };

    Config.prototype.set_master_ring = function(r, cb) {
      var err, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      this._master_ring = r;
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/node-installer/src/config.iced",
            funcname: "Config.set_master_ring"
          });
          _this._master_ring.index(__iced_deferrals.defer({
            assign_fn: (function(__slot_1) {
              return function() {
                err = arguments[0];
                return __slot_1._keyring_index = arguments[1];
              };
            })(_this),
            lineno: 79
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          return cb(err);
        };
      })(this));
    };

    Config.prototype.url_prefix = function() {
      var prot, u;
      if ((u = this.argv.get("u", "url-prefix"))) {
        return u;
      } else {
        prot = this.argv.get("S", "no-https") ? 'http' : 'https';
        return constants.url_prefix[prot];
      }
    };

    Config.prototype.install_prefix = function() {
      return process.env.PREFIX || this.argv.get("p", "prefix");
    };

    Config.prototype.make_url = function(u) {
      return url_join(this.url_prefix(), u);
    };

    Config.prototype.get_tmpdir = function() {
      return this._tmpdir;
    };

    Config.prototype.get_no_gpg_options = function() {
      return this.argv.get("O", "no-gpg-options");
    };

    Config.prototype.make_tmpdir = function(cb) {
      var err, p, r, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      err = null;
      (function(_this) {
        return (function(__iced_k) {
          if (_this._tmpdir == null) {
            r = base64u.encode(prng(16));
            p = kpath.split(tmpdir()).concat(["keybase_install_" + r]);
            _this._tmpdir = kpath.unsplit(p);
            (function(__iced_k) {
              __iced_deferrals = new iced.Deferrals(__iced_k, {
                parent: ___iced_passed_deferral,
                filename: "/home/max/src/keybase/node-installer/src/config.iced",
                funcname: "Config.make_tmpdir"
              });
              fs.mkdir(_this._tmpdir, 0x1c0, __iced_deferrals.defer({
                assign_fn: (function() {
                  return function() {
                    return err = arguments[0];
                  };
                })(),
                lineno: 111
              }));
              __iced_deferrals._fulfill();
            })(function() {
              return __iced_k(log.info("Made temporary directory: " + _this._tmpdir));
            });
          } else {
            return __iced_k();
          }
        });
      })(this)((function(_this) {
        return function() {
          return cb(err);
        };
      })(this));
    };

    Config.prototype.cleanup = function(cb) {
      var esc, f, files, p, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      esc = make_esc(cb, "Installer::cleanup");
      log.debug("+ cleanup " + this._tmpdir);
      (function(_this) {
        return (function(__iced_k) {
          if (_this._tmpdir == null) {
            return __iced_k();
          } else {
            (function(__iced_k) {
              if (_this.argv.get("C", "skip-cleanup")) {
                return __iced_k(log.info("Preserving tmpdir " + _this._tmpdir + " as per command-line switch"));
              } else {
                log.info("cleaning up tmpdir " + _this._tmpdir);
                (function(__iced_k) {
                  __iced_deferrals = new iced.Deferrals(__iced_k, {
                    parent: ___iced_passed_deferral,
                    filename: "/home/max/src/keybase/node-installer/src/config.iced",
                    funcname: "Config.cleanup"
                  });
                  fs.readdir(_this._tmpdir, esc(__iced_deferrals.defer({
                    assign_fn: (function() {
                      return function() {
                        return files = arguments[0];
                      };
                    })(),
                    lineno: 125
                  })));
                  __iced_deferrals._fulfill();
                })(function() {
                  (function(__iced_k) {
                    var _i, _len, _ref3, _results, _while;
                    _ref3 = files;
                    _len = _ref3.length;
                    _i = 0;
                    _results = [];
                    _while = function(__iced_k) {
                      var _break, _continue, _next;
                      _break = function() {
                        return __iced_k(_results);
                      };
                      _continue = function() {
                        return iced.trampoline(function() {
                          ++_i;
                          return _while(__iced_k);
                        });
                      };
                      _next = function(__iced_next_arg) {
                        _results.push(__iced_next_arg);
                        return _continue();
                      };
                      if (!(_i < _len)) {
                        return _break();
                      } else {
                        f = _ref3[_i];
                        p = kpath.join(_this._tmpdir, f);
                        log.debug("| Unlink " + p);
                        (function(__iced_k) {
                          __iced_deferrals = new iced.Deferrals(__iced_k, {
                            parent: ___iced_passed_deferral,
                            filename: "/home/max/src/keybase/node-installer/src/config.iced",
                            funcname: "Config.cleanup"
                          });
                          fs.unlink(p, esc(__iced_deferrals.defer({
                            lineno: 129
                          })));
                          __iced_deferrals._fulfill();
                        })(_next);
                      }
                    };
                    _while(__iced_k);
                  })(function() {
                    (function(__iced_k) {
                      __iced_deferrals = new iced.Deferrals(__iced_k, {
                        parent: ___iced_passed_deferral,
                        filename: "/home/max/src/keybase/node-installer/src/config.iced",
                        funcname: "Config.cleanup"
                      });
                      fs.rmdir(_this._tmpdir, esc(__iced_deferrals.defer({
                        lineno: 130
                      })));
                      __iced_deferrals._fulfill();
                    })(__iced_k);
                  });
                });
              }
            })(__iced_k);
          }
        });
      })(this)((function(_this) {
        return function() {
          log.debug("- cleanup " + _this._tmpdir + " -> OK");
          return cb(null);
        };
      })(this));
    };

    Config.prototype.get_proxy = function() {
      return this.argv.get("x", "proxy") || process.env.https_proxy || process.env.http_proxy;
    };

    Config.prototype.request = function(u, cb) {
      var body, err, opts, res, url, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      url = u.match("^https?://") ? u : this.make_url(u);
      opts = {
        url: url,
        headers: {
          "X-Keybase-Installer": fullname()
        },
        maxRedirects: 10,
        progress: 50000,
        proxy: this.get_proxy()
      };
      log.info("Fetching URL " + url);
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/node-installer/src/config.iced",
            funcname: "Config.request"
          });
          request(opts, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                res = arguments[1];
                return body = arguments[2];
              };
            })(),
            lineno: 150
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          log.debug(" * fetched -> " + (typeof res !== "undefined" && res !== null ? res.statusCode : void 0));
          return cb(err, res, body);
        };
      })(this));
    };

    Config.prototype.set_keys = function(keys) {
      if ((this._keys == null) || this._keys.version !== keys.version) {
        log.info("Using keyset version v" + keys.version);
        return this._keys = keys;
      }
    };

    Config.prototype.set_alt_cmds = function() {
      var c, n;
      if ((c = this.argv.get("g", "gpg")) != null) {
        this._alt_cmds.gpg = c;
        set_gpg_cmd(c);
      }
      if ((n = this.argv.get("n", "npm")) != null) {
        return this._alt_cmds.npm = n;
      }
    };

    Config.prototype.get_alt_cmd = function(k) {
      return this._alt_cmds[k];
    };

    Config.prototype.get_cmd = function(k) {
      return this.get_alt_cmd(k) || k;
    };

    Config.prototype.key_version = function() {
      return this._keys.version;
    };

    Config.prototype.set_index = function(i) {
      return this._index = i;
    };

    Config.prototype.index = function() {
      return this._index;
    };

    Config.prototype.index_lookup_hash = function(v) {
      var _ref3, _ref4;
      return (_ref3 = this._index["package"]) != null ? (_ref4 = _ref3.all) != null ? _ref4[v] : void 0 : void 0;
    };

    Config.prototype.oneshot_verify = function(_arg, cb) {
      var err, file, json, query, sig, which, ___iced_passed_deferral, __iced_deferrals, __iced_k;
      __iced_k = __iced_k_noop;
      ___iced_passed_deferral = iced.findDeferral(arguments);
      which = _arg.which, sig = _arg.sig, file = _arg.file;
      query = this._keys[which].fingerprint();
      (function(_this) {
        return (function(__iced_k) {
          __iced_deferrals = new iced.Deferrals(__iced_k, {
            parent: ___iced_passed_deferral,
            filename: "/home/max/src/keybase/node-installer/src/config.iced",
            funcname: "Config.oneshot_verify"
          });
          _this.master_ring().oneshot_verify({
            query: query,
            file: file,
            sig: sig,
            single: true
          }, __iced_deferrals.defer({
            assign_fn: (function() {
              return function() {
                err = arguments[0];
                return json = arguments[1];
              };
            })(),
            lineno: 192
          }));
          __iced_deferrals._fulfill();
        });
      })(this)((function(_this) {
        return function() {
          return cb(err, json);
        };
      })(this));
    };

    return Config;

  })();

}).call(this);
